### PR TITLE
feat(tt-5): setup s3 uploading from an admin

### PR DESCRIPTION
**Pull request by joelhooks**
https://bobbyhadz.com/blog/aws-s3-presigned-url-react - this is a pretty good approach and had some code to yoink to get us a bit down the road.

Now, assuming you have the appropriate roles, you can visit /admin and there is a file chooser and upload button. It is filtered for video files and will upload the selected file to s3 and give you the resulting public URL

next steps would be to create a video resource from that URL in sanity, but that's a little more complex and this will be handy for now

signed

### CODE REVIEW

File: apps/total-typescript/src/server/get-middleware-response.ts [hunk 1]
```diff
@@ -42,6 +46,18 @@ export async function getMiddlewareResponse(req: NextRequest) {
...
+    try {
+      const user = UserSchema.parse(token)
+      const ability = getCurrentAbility({user})
+      if (!ability.can('create', 'Content')) {
+        response = rewriteToPath('/login', req)
+      }
+    } catch (error) {
+      response = rewriteToPath('/login', req)
+    }
+  }
```
Comment (joelhooks) [inline apps/total-typescript/src/server/get-middleware-response.ts:59]: this is guarding the /admin route

File: apps/total-typescript/src/module-builder/upload-file.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  const presignedPostUrl = await getPresignedPostUrl(
+    fileType,
+    fileContents.name,
+  )
+
+  console.log('presignedPostUrl', presignedPostUrl)
+
+  await axios.put(presignedPostUrl.signedUrl, fileContents, {
+    headers: {'Content-Type': 'application/octet-stream'},
+    onUploadProgress,
```
Comment (joelhooks) [inline apps/total-typescript/src/module-builder/upload-file.ts:25]: axios has `onUploadProgress` because it uses xhr under the hood. You can't get progress with fetch.

File: apps/total-typescript/src/module-builder/upload-file.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  // rebuild it as a fresh new thing
+  return `${scrubbed}.${fileExtension(fullFilename)}`
+}
+
+async function getPresignedPostUrl(fileType: string, fileName: string) {
+  const {data: presignedPostUrl} = await axios.get<PresignedPostUrlResponse>(
+    `${SIGNING_URL}?contentType=${fileType}&objectName=${prepareFileName(
+      fileName,
+    )}`,
+  )
```
Comment (joelhooks) [inline apps/total-typescript/src/module-builder/upload-file.ts:54]: api endpoint that creates the s3 signed put url

File: apps/total-typescript/src/module-builder/use-file-change.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  fileContents: null,
+}
+
+function bytesToMb(bytes: number) {
+  const mb = bytes / 1000000
+
+  return mb
+}
+
+export function useFileChange() {
```
Comment (joelhooks) [inline apps/total-typescript/src/module-builder/use-file-change.ts:17]: managing the state of the file input

File: apps/total-typescript/src/offer/survey/survey-question.tsx [hunk 5]
```diff
@@ -32,7 +32,7 @@ export type SurveyConfig = {
   questionBodyRenderer?: (question: any) => void
 }
 
-export const SurveyQuestionContext =
+export const SurveyQuestionContext: React.Context<InternalQuestionContextValue> =
```
Comment (Creeland) [inline apps/total-typescript/src/offer/survey/survey-question.tsx:35]: This change looks unrelated

File: apps/total-typescript/src/server/get-middleware-response.ts (hunk 1 above)
Comment (Creeland) [reply apps/total-typescript/src/server/get-middleware-response.ts:40]: I haven't used middleware in Next before so it's neat to see it in action

**Review by Creeland — APPROVED** [review]
